featurebyte/service/relationship.py

```diff
@@ -138,15 +138,65 @@ async def add_relationship(self, parent: ParentT, child_id: ObjectId) -> Relatio
 
             return updated_document
 
-    @staticmethod
-    def _validate_remove_relationship_operation(
-        parent_obj: Relationship, child_obj: Relationship
+    async def _remove_ancestor_ids(
+        self,
+        document: Relationship,
+        remove_parent_id: Optional[ObjectId],
+        remove_ancestor_ids: Set[ObjectId],
+        checked_descendant_ids: Set[ObjectId],
     ) -> None:
-        has_relationship = [par for par in child_obj.parents if par.id == parent_obj.id]
-        if not has_relationship:
-            raise DocumentUpdateError(
-                f'Object "{parent_obj.name}" is not the parent of object "{child_obj.name}".'
+        """
+        Remove ancestor IDs from the document and update all descendant objects
+
+        Parameters
+        ----------
+        document: Relationship
+            Document object
+        remove_parent_id: Optional[ObjectId]
+            Parent ID to be removed
+        remove_ancestor_ids: Set[ObjectId]
+            Ancestor IDs to be removed
+        checked_descendant_ids: Set[ObjectId]
+            Descendant IDs that have been checked
+        """
+        if document.id in checked_descendant_ids:
+            return
+
+        # find the ancestor_ids of the parent objects & keep them
+        keep_parents = [par for par in document.parents if par.id != remove_parent_id]
+        keep_parent_ids = [par.id for par in keep_parents]
+
+        keep_ancestor_ids = set(keep_parent_ids)
+        if keep_parent_ids:
+            async for obj in self.document_service.list_documents_as_dict_iterator(
+                query_filter={"_id": {"$in": keep_parent_ids}},
+                projection={"ancestor_ids": 1},
+            ):
+                keep_ancestor_ids.update(obj["ancestor_ids"])
+
+        ancestor_ids = (
+            set(document.ancestor_ids).difference(remove_ancestor_ids).union(keep_ancestor_ids)
+        )
+        await self.document_service.update_document(
+            document_id=document.id,
+            data=self.prepare_document_update_payload(
+                ancestor_ids=ancestor_ids,
+                parents=keep_parents,
+            ),
+        )
+        checked_descendant_ids.add(document.id)
+
+        async for descendant_obj in self.document_service.list_documents_iterator(
+            query_filter={"parents.id": document.id},
+        ):
+            await self._remove_ancestor_ids(
```

Comment:
Curious what is the expected number of queries this function will call since it's recursive?

Reply:
We don't expect the number to be high. It is scaled with number of descendent of a particular entity where its parent get removed.

Reply:
Before current changes, there are like 2 update queries
* 1 update current parent-child entity relationship
* 1 update descendent of the current child entity (to update descendants' ancestor IDs)

After current changes, the number of queries is scaled with number of descendants that the parent entity has.

Reply:
If we change the filter condition to `{"ancestor_ids": document.id}` then this function does not need to be recursive?

Reply:
Do you mean making multiple updates inside the non-recursive function?

Reply:
Looks like we need to do this recursively since there is no restriction on shared ancestry from different parents